sdk/servicebus/azservicebus/receiver.go

```diff
@@ -14,11 +24,306 @@ const (
 	SubQueueTransfer   = "transferDeadLetter"
 )
 
-type ServiceBusReceiver struct{}
+type Receiver struct {
+	config struct {
+		ReceiveMode    ReceiveMode
+		FullEntityPath string
+		Entity         entity
+
+		RetryOptions struct {
+			Times int
+			Delay time.Duration
+		}
+	}
+
+	mu       sync.Mutex
+	ns       legacyNamespace
+	receiver internal.LegacyReceiver
+
+	linkState *linkState
+}
+
+type ReceiverOption func(receiver *Receiver) error
+
+// ReceiverWithSubQueue allows you to open the sub queue (ie: dead letter queues, transfer dead letter queues)
+// for a queue or subscription.
+func ReceiverWithSubQueue(subQueue SubQueue) ReceiverOption {
+	return func(receiver *Receiver) error {
+		switch subQueue {
+		case SubQueueDeadLetter:
+		case SubQueueTransfer:
+		case "":
+			receiver.config.Entity.Subqueue = subQueue
+		default:
+			return fmt.Errorf("unknown SubQueue %s", subQueue)
+		}
+
+		return nil
+	}
+}
+
+// ReceiverWithReceiveMode controls the receive mode for the receiver.
+func ReceiverWithReceiveMode(receiveMode ReceiveMode) ReceiverOption {
+	return func(receiver *Receiver) error {
+		if receiveMode != PeekLock && receiveMode != ReceiveAndDelete {
+			return fmt.Errorf("invalid receive mode specified %d", receiveMode)
+		}
+
+		receiver.config.ReceiveMode = receiveMode
+		return nil
+	}
+}
+
+func ReceiverWithQueue(queue string) ReceiverOption {
+	return func(receiver *Receiver) error {
+		receiver.config.Entity.Queue = queue
+		return nil
+	}
+}
+
+func ReceiverWithSubscription(topic string, subscription string) ReceiverOption {
+	return func(receiver *Receiver) error {
+		receiver.config.Entity.Topic = topic
+		receiver.config.Entity.Subscription = subscription
+		return nil
+	}
+}
+
+func newReceiver(ns legacyNamespace, options ...ReceiverOption) (*Receiver, error) {
+	receiver := &Receiver{
+		config: struct {
+			ReceiveMode    ReceiveMode
+			FullEntityPath string
+			Entity         entity
+			RetryOptions   struct {
+				Times int
+				Delay time.Duration
+			}
+		}{
+			ReceiveMode: PeekLock,
+		},
+		ns:        ns,
+		linkState: newLinkState(context.Background(), errClosed{link: "receiver"}),
+	}
+
+	for _, opt := range options {
+		if err := opt(receiver); err != nil {
+			return nil, err
+		}
+	}
+
+	entityPath, err := receiver.config.Entity.String()
+
+	if err != nil {
+		return nil, err
+	}
+
+	receiver.config.FullEntityPath = entityPath
+	return receiver, nil
+}
+
+type ReceiveOptions struct {
```

Comment:
It's unfortunate that this has to be exported.  An option to get rid of this, although it's a bit more code, is to have `ReceiveOption` be an interface with an internal `apply()` method.  That way you don't have to export any of the dependent types.

```go
type ReceiveOption interface {
	apply(*receiveOption)
}

type receiveOption struct {
	maxWaitTime time.Duration
}

type receiveOptionFn func(*receiveOption)

func (r receiveOptionFn) apply(ro *receiveOption) {
	r(ro)
}

func ReceiveWithMaxWaitTime(max time.Duration) ReceiveOption {
	return receiveOptionFn(func(ro *receiveOption) {
		ro.maxWaitTime = max
	})
}
```

Reply:
Also, and we don't have to change this now, but if we make `apply()` (or whatever impl we go with) return an `error` then it gives us future flexibility if we find ourselves in the situation where we want to validate a config option.

Reply:
Seems sensible to me. I'm assuming we're going to have this same problem in autogenerated code. What pattern are we laying out there today (the apply version or the version I have above)?